fuse_constraints/src/marginalize_variables.cpp

```diff
@@ -321,20 +327,28 @@ LinearTerm linearize(
                              "during the jacobian computation.");
   }
 
-  // Update the jacobians with the local parameterizations.
+  // Update the jacobians with the local parameterizations. This potentially changes the size of the Jacobian block.
+  // The classic example is a quaternion parameter (4 components) only has 3 degrees of freedom. So the Jacobian will
+  // be transformed from 4 columns to 3 columns after the local parameterization is applied.
+  // We also check for variables that have been marked as constants. Since these variables cannot change value, their
+  // derivatives/Jacobians should be zero.
   for (size_t index = 0ul; index < variable_count; ++index)
   {
     const auto& variable_uuid = variable_uuids[index];
     const auto& variable = graph.getVariable(variable_uuid);
     auto local_parameterization = variable.localParameterization();
+    auto& jacobian = result.A[index];
     if (local_parameterization)
     {
-      auto& jacobian = result.A[index];
       fuse_core::MatrixXd J(local_parameterization->GlobalSize(), local_parameterization->LocalSize());
       local_parameterization->ComputeJacobian(variable_values[index], J.data());
       delete local_parameterization;
       jacobian *= J;
     }
+    if (variable.holdConstant())
+    {
+      jacobian.setZero();
+    }
```

Comment:
Shouldn't this be checked before "if (local_parameterization)"? If it is hold constant, any need to compute the jacobian for this variable?

Reply:
Yes, it is wasteful. And yes, that does bother me.

The problem is the local parameterization potentially changes the size the Jacobian matrix. By applying the local parameterization even for constant variables, we ensure the block of zeros is the correct size.

But since you bring it up, we could do something like:
```
if (variable.holdConstant())
{
  if (local_parameterization)
  {
    jacobian.resize(jacobian.rows(), local_parameterization->LocalSize());
  }
  jacobian.setZero();
}
else if (local_parameterization)
{
  ...
}
```
Let me see if I can make that work without too much trouble.

Reply:
There's even a fancy `Eigen::NoChange` option so I don't need to look up the row count.
```
jacobian.resize(Eigen::NoChange, local_parameterization->LocalSize());
```